### PR TITLE
Happy Eyeballs (IP6 fallback to IP4) implementation of http client.

### PR DESCRIPTION
**Pull request by BaronGreenback**
Implementation of the default http client implementing happy eyeballs. https://github.com/jellyfin/jellyfin/issues/1002

Default implementation now uses happy eyeballs. (IP6 with IP4 fallback).

1) This implementation will cease using IPv6 in case of exception. 
If this exception is due to physical network changes NetworkManager will reset the IP6 capability, which will re-enable the functionality.

2) IP6 functionality is controlled via the network.xml. If, for some reason, NetworkManager hasn't been instantiated, IPv4 will be used. (thinking of testing)

3) This implementation is ignored by the DLNA sub-system, which continues to use single attempt due to it accessing IP addresses.

### CODE REVIEW

File: Jellyfin.Server/Startup.cs [hunk 1]
```diff
@@ -67,15 +70,21 @@ public void ConfigureServices(IServiceCollection services)
...
                     c.DefaultRequestHeaders.Accept.Add(acceptJsonHeader);
                     c.DefaultRequestHeaders.Accept.Add(acceptXmlHeader);
                     c.DefaultRequestHeaders.Accept.Add(acceptAnyHeader);
                 })
-                .ConfigurePrimaryHttpMessageHandler(x => new DefaultHttpClientHandler());
+                .ConfigurePrimaryHttpMessageHandler(x =>
+                {
+                    return new SocketsHttpHandler
+                    {
+                        AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate,
```
Comment (crobibero) [inline Jellyfin.Server/Startup.cs:84]:
```suggestion
                        AutomaticDecompression = DecompressionMethods.All,
```

**Review by Ullmie02 — APPROVED** [review]

File: Jellyfin.Server/Startup.cs [hunk 2]
```diff
@@ -96,6 +98,15 @@ public void ConfigureServices(IServiceCollection services)
                 })
                 .ConfigurePrimaryHttpMessageHandler(defaultHttpClientHandlerDelegate);
 
+            services.AddHttpClient(NamedClient.DirectIp, c =>
```
Comment (crobibero) [inline Jellyfin.Server/Startup.cs:101]: What's the difference between `Default` and `DirectIp`? Right now they seem to be the same

Comment (BaronGreenback) [reply Jellyfin.Server/Startup.cs:107]: just updated after bond's change.

**Review by crobibero — APPROVED** [review]

File: Jellyfin.Networking/HappyEyeballs/HttpClientExtension.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+        /// <summary>
+        /// Gets a value indicating whether the initial IPv6 check has been performed (to determine whether v6 is available or not).
+        /// </summary>
+        private static bool _hasResolvedIPv6Availability;
+
+        /// <summary>
+        /// Gets or sets a value indicating whether IPv6 should be preferred. Value may change based on runtime failures.
+        /// </summary>
+        public static bool? UseIPv6 { get; set; } = null;
```
Comment (cvium) [inline Jellyfin.Networking/HappyEyeballs/HttpClientExtension.cs:48]: I don't know why I didn't notice it sooner, but this is not thread safe.

**Review by cvium — CHANGES_REQUESTED** [review]
We should fix the thread safety issue

**Comment by Yash-Garg** [comment]
Any updates on this? Would like to try it out if this works.

**Comment by AkechiShiro** [comment]
Hey any update on this so far, looks like it's been abandoned ?

**Comment by Shadowghost** [comment]
Superseded by #8598

**Pull request closed without merging**